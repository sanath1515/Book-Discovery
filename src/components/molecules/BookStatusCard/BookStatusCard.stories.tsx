import React from "react";
import BookStatusCard from "./BookStatusCard";

export default {
  title: "molecules/BookStatusCard",
  component: BookStatusCard,
};

const Template = () => (
  <BookStatusCard
    title="Inorganic Chemistry"
    authorName="Joe Person"
    field="Chemistry"
    image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/B4303A95-C174-4B8B-8F27-D9911CAA7D09.png"
    upCount={16}
    total={100}
    id="1"
  />
);

export const Default = Template.bind({});
