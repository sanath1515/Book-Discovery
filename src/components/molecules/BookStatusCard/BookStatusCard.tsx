import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import { CardActions, Typography } from "@material-ui/core";
import { COLORS } from "../../../theme/constants";
import theme from "../../../theme/theme";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router";

const useStyles = makeStyles({
  root: {
    width: "255px",
    height: "108px",
    padding: theme.spacing(2, 0, 2, 1),
    borderRadius: "4px",
    border: "solid 1px",
    borderColor: COLORS.GREY_300,
  },
  image: {
    width: "80px",
    height: "105px",
    margin: theme.spacing(1),
    objectFit: "fill",
  },
  title: {
    width: "152px",
    height: "22px",
    color: COLORS.CHARCOAL_GREY,
    textAlign: "left",
    paddingLeft: theme.spacing(1),
  },
  author: {
    width: "20px",
    height: "18px",
    color: COLORS.ICON_500,
    paddingLeft: theme.spacing(0.5),
  },
  authorName: {
    width: "160px",
    height: "22px",
    paddingTop: theme.spacing(1.5),
    color: COLORS.CHARCOAL_GREY,
    textAlign: "left",
  },
  field: {
    width: "55px",
    height: "18px",
    textTransform: "none",
    color: COLORS.ICON_500,
  },
  fieldName: {
    width: "64px",
    height: "22px",
    textTransform: "none",
    color: COLORS.CHARCOAL_GREY,
    textAlign: "left",
  },
  pages: {
    width: "auto",
    height: "18px",
    textTransform: "none",
    color: COLORS.CHARCOAL_GREY_100,
  },
  actions: {
    padding: theme.spacing(0.5),
    margin: theme.spacing(0, 0.5),
  },
  actionArea: {
    display: "flex",
    width: "auto",
  },
  content: {
    padding: theme.spacing(0),
  },
});
export type BookStatusCardProps = {
  id: string;
  title: string;
  authorName: string;
  field: string;
  image: string;
  upCount?: number;
  total?: number;
};
const BookStatusCard: React.FC<BookStatusCardProps> = ({
  id,
  title,
  field,
  authorName,
  image,
  upCount,
  total,
}) => {
  const classes = useStyles();

  const dispatch = useDispatch();
  const history = useHistory();

  const handleClick = (bookId: string) => {
    dispatch({ type: "SET_SINGLE_VIEW_ID", payload: bookId });
    history.push("/bookdetail");
  };
  return (
    <Card className={classes.root}>
      <CardActionArea className={classes.actionArea}>
        <CardMedia
          className={classes.image}
          component="img"
          image={image}
          onClick={() => handleClick(id)}
          data-testid="card"
        />
        <CardContent className={classes.content}>
          <Typography className={classes.title} variant="subtitle1">
            {title}
          </Typography>
          <CardActions className={classes.actions}>
            <Typography className={classes.author} variant="body2">
              by
            </Typography>
            <Typography className={classes.authorName} variant="body2">
              {authorName}
            </Typography>
          </CardActions>
          <CardActions className={classes.actions}>
            <Typography className={classes.field} variant="overline">
              Category:
            </Typography>
            <Typography className={classes.fieldName} variant="body2">
              {field}
            </Typography>
          </CardActions>
          <CardActions className={classes.actions}>
            <Typography className={classes.pages} variant="overline">
              {upCount} left / {total} pages
            </Typography>
          </CardActions>
        </CardContent>
      </CardActionArea>
    </Card>
  );
};

export default BookStatusCard;
