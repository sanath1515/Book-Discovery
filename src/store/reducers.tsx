export type bookState = {
    searchText:string,
    category:string,
    singleViewId:string
}

export const initialBookState = {
    searchText:"",
    category:"",
    singleViewId:"1"
}

type Action = {type:string,payload:string}

export const booksReducer = (state:bookState = initialBookState,action:Action)=>
{
    switch(action.type)
    {
        case "SET_CATEGORY":{
            return {...state,category:action.payload};
        }
        case "SET_SEARCH_TEXT":{
            return {...state,searchText:action.payload};
        }
        case "SET_SINGLE_VIEW_ID":{
            return {...state,singleViewId:action.payload};
        }
        default:
            return state;
    }
}