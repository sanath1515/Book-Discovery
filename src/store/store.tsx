import { createStore } from 'redux';
import { booksReducer } from './reducers';

const store = createStore(booksReducer);

export default store;