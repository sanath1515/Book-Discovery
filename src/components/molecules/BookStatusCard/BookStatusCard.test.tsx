import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import BookStatusCard from "./BookStatusCard";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import store from "../../../store/store";


afterEach(cleanup);

describe("Book Status Card Testing", () => {
  it("Matches Snapshot", () => {
    const { container } = render(
      <Router>
      <Provider store={store}>
      <BookStatusCard
        title="Nature Chemistry"
        field="Chemistry"
        authorName="J D Lee"
        image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
        upCount={12}
        total={200}
        id="1"
      />
      </Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });

  it("Checking The title", () => {
    render(
      <Router>
      <Provider store={store}>
      <BookStatusCard
        title="Nature Chemistry"
        field="Chemistry"
        authorName="J D Lee"
        image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
        upCount={12}
        total={200}
        id="1"
      />
      </Provider>
      </Router>
    );

    expect(screen.getByText("Nature Chemistry")).toBeInTheDocument();
    expect(screen.getByText("J D Lee")).toBeInTheDocument();
    expect(screen.getByText("Chemistry")).toBeInTheDocument();
  });
  it("Should click correctly", () => {
    const { getByTestId } = render(
      <Router>
        <Provider store={store}>
        <BookStatusCard
        title="Nature Chemistry"
        field="Chemistry"
        authorName="J D Lee"
        image="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
        upCount={12}
        total={200}
        id="1"
      />
        </Provider>
      </Router>
    );
    fireEvent.click(getByTestId("card"));
    expect(fireEvent.click(getByTestId("card"))).toBeTruthy();
  });
});
