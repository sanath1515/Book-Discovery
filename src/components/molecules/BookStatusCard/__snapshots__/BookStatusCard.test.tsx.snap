// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Status Card Testing Matches Snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root makeStyles-actionArea-10"
      tabindex="0"
      type="button"
    >
      <img
        class="MuiCardMedia-root makeStyles-image-2 MuiCardMedia-media MuiCardMedia-img"
        data-testid="card"
        src="https://cdn.zeplin.io/5f03041d8c8fc11335546618/assets/920F4926-3BD5-47FB-A32A-32C03DE1BB71.png"
      />
      <div
        class="MuiCardContent-root makeStyles-content-11"
      >
        <h6
          class="MuiTypography-root makeStyles-title-3 MuiTypography-subtitle1"
        >
          Nature Chemistry
        </h6>
        <div
          class="MuiCardActions-root makeStyles-actions-9 MuiCardActions-spacing"
        >
          <p
            class="MuiTypography-root makeStyles-author-4 MuiTypography-body2"
          >
            by
          </p>
          <p
            class="MuiTypography-root makeStyles-authorName-5 MuiTypography-body2"
          >
            J D Lee
          </p>
        </div>
        <div
          class="MuiCardActions-root makeStyles-actions-9 MuiCardActions-spacing"
        >
          <span
            class="MuiTypography-root makeStyles-field-6 MuiTypography-overline"
          >
            Category:
          </span>
          <p
            class="MuiTypography-root makeStyles-fieldName-7 MuiTypography-body2"
          >
            Chemistry
          </p>
        </div>
        <div
          class="MuiCardActions-root makeStyles-actions-9 MuiCardActions-spacing"
        >
          <span
            class="MuiTypography-root makeStyles-pages-8 MuiTypography-overline"
          >
            12
             left / 
            200
             pages
          </span>
        </div>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
